File: Python Playground/12_graph_dfs_bfs.py
"""
GRAPH, DFS & BFS - Complete Guide for Interview Preparation
============================================================

CORE CONCEPTS:
--------------
GRAPH REPRESENTATION:
1. Adjacency Matrix: 2D array, O(V²) space, O(1) edge check
2. Adjacency List: dict/list of lists, O(V+E) space, O(deg(v)) edge check
3. Edge List: list of (u, v) pairs

GRAPH TYPES:
- Directed vs Undirected
- Weighted vs Unweighted
- Cyclic vs Acyclic (DAG)
- Connected vs Disconnected

DFS (Depth-First Search):
- Uses stack (recursion or explicit)
- Time: O(V+E), Space: O(V)
- Use: cycle detection, topological sort, connected components

BFS (Breadth-First Search):
- Uses queue
- Time: O(V+E), Space: O(V)
- Use: shortest path (unweighted), level-order traversal

TRICKY PARTS:
-------------
1. Graph can be implicit (generate neighbors on the fly)
2. Visited tracking crucial to avoid infinite loops
3. Directed vs undirected affects algorithm
4. DFS can hit stack limit - use iterative if deep
5. Disconnected graphs need to iterate all nodes

COMMON PATTERNS:
----------------
1. DFS: Connected components, cycle detection, path finding
2. BFS: Shortest path, level-order
3. Union Find: Connected components, cycle detection
4. Topological Sort: DAG ordering
5. Shortest Path: Dijkstra, Bellman-Ford
"""

from typing import List, Set, Dict, Optional
from collections import deque, defaultdict
import heapq


# ============================================================================
# PATTERN 1: DFS - RECURSIVE
# ============================================================================

def dfs_recursive(graph: Dict[int, List[int]], start: int, visited: Set[int]):
    """
    Basic DFS traversal - recursive.

    Time: O(V+E), Space: O(V) for recursion stack
    """
    visited.add(start)
    print(start, end=' ')

    for neighbor in graph.get(start, []):
        if neighbor not in visited:
            dfs_recursive(graph, neighbor, visited)


def dfs_iterative(graph: Dict[int, List[int]], start: int) -> List[int]:
    """
    DFS traversal - iterative using stack.

    Time: O(V+E), Space: O(V)
    """
    visited = set()
    stack = [start]
    result = []

    while stack:
        node = stack.pop()

        if node not in visited:
            visited.add(node)
            result.append(node)

            # Add neighbors in reverse to maintain order
            for neighbor in reversed(graph.get(node, [])):
                if neighbor not in visited:
                    stack.append(neighbor)

    return result


# ============================================================================
# PATTERN 2: BFS
# ============================================================================

def bfs(graph: Dict[int, List[int]], start: int) -> List[int]:
    """
    BFS traversal using queue.

    Time: O(V+E), Space: O(V)
    """
    visited = {start}
    queue = deque([start])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)

        for neighbor in graph.get(node, []):
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)

    return result


def bfs_level_order(graph: Dict[int, List[int]], start: int) -> List[List[int]]:
    """
    BFS with level tracking.

    Time: O(V+E), Space: O(V)
    """
    visited = {start}
    queue = deque([start])
    result = []

    while queue:
        level = []
        level_size = len(queue)

        for _ in range(level_size):
            node = queue.popleft()
            level.append(node)

            for neighbor in graph.get(node, []):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

        result.append(level)

    return result


# ============================================================================
# PATTERN 3: NUMBER OF ISLANDS / CONNECTED COMPONENTS
# ============================================================================

def num_islands(grid: List[List[str]]) -> int:
    """
    LeetCode 200 - Number of islands (connected components).

    Time: O(m*n), Space: O(m*n) for recursion
    """
    if not grid:
        return 0

    m, n = len(grid), len(grid[0])
    count = 0

    def dfs(i, j):
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != '1':
            return

        grid[i][j] = '0'  # Mark as visited

        # Explore 4 directions
        dfs(i + 1, j)
        dfs(i - 1, j)
        dfs(i, j + 1)
        dfs(i, j - 1)

    for i in range(m):
        for j in range(n):
            if grid[i][j] == '1':
                dfs(i, j)
                count += 1

    return count


def max_area_island(grid: List[List[int]]) -> int:
    """
    LeetCode 695 - Max area of island.

    Time: O(m*n), Space: O(m*n)
    """
    if not grid:
        return 0

    m, n = len(grid), len(grid[0])

    def dfs(i, j):
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != 1:
            return 0

        grid[i][j] = 0  # Mark visited

        return 1 + dfs(i+1, j) + dfs(i-1, j) + dfs(i, j+1) + dfs(i, j-1)

    max_area = 0

    for i in range(m):
        for j in range(n):
            if grid[i][j] == 1:
                max_area = max(max_area, dfs(i, j))

    return max_area


def count_connected_components(n: int, edges: List[List[int]]) -> int:
    """
    LeetCode 323 - Number of connected components in undirected graph.

    Time: O(V+E), Space: O(V+E)
    """
    # Build adjacency list
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    visited = set()
    count = 0

    def dfs(node):
        visited.add(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                dfs(neighbor)

    for node in range(n):
        if node not in visited:
            dfs(node)
            count += 1

    return count


# ============================================================================
# PATTERN 4: CYCLE DETECTION
# ============================================================================

def has_cycle_undirected(n: int, edges: List[List[int]]) -> bool:
    """
    Detect cycle in undirected graph using DFS.

    Time: O(V+E), Space: O(V)
    """
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    visited = set()

    def dfs(node, parent):
        visited.add(node)

        for neighbor in graph[node]:
            if neighbor not in visited:
                if dfs(neighbor, node):
                    return True
            elif neighbor != parent:
                # Visited and not parent = cycle
                return True

        return False

    for node in range(n):
        if node not in visited:
            if dfs(node, -1):
                return True

    return False


def has_cycle_directed(n: int, edges: List[List[int]]) -> bool:
    """
    LeetCode 207 variation - Detect cycle in directed graph.

    Time: O(V+E), Space: O(V)
    Tricky: Need to track nodes in current path
    """
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    UNVISITED, VISITING, VISITED = 0, 1, 2
    state = [UNVISITED] * n

    def dfs(node):
        if state[node] == VISITING:
            return True  # Cycle detected
        if state[node] == VISITED:
            return False

        state[node] = VISITING

        for neighbor in graph[node]:
            if dfs(neighbor):
                return True

        state[node] = VISITED
        return False

    for node in range(n):
        if dfs(node):
            return True

    return False


# ============================================================================
# PATTERN 5: SHORTEST PATH (BFS - Unweighted)
# ============================================================================

def shortest_path_binary_matrix(grid: List[List[int]]) -> int:
    """
    LeetCode 1091 - Shortest path in binary matrix.

    Time: O(n²), Space: O(n²)
    """
    n = len(grid)

    if grid[0][0] == 1 or grid[n-1][n-1] == 1:
        return -1

    # 8 directions
    directions = [(0,1), (1,0), (0,-1), (-1,0), (1,1), (1,-1), (-1,1), (-1,-1)]

    queue = deque([(0, 0, 1)])  # (row, col, distance)
    grid[0][0] = 1  # Mark visited

    while queue:
        row, col, dist = queue.popleft()

        if row == n-1 and col == n-1:
            return dist

        for dr, dc in directions:
            r, c = row + dr, col + dc

            if 0 <= r < n and 0 <= c < n and grid[r][c] == 0:
                grid[r][c] = 1  # Mark visited
                queue.append((r, c, dist + 1))

    return -1


def walls_and_gates(rooms: List[List[int]]) -> None:
    """
    LeetCode 286 - Walls and gates (multi-source BFS).

    Time: O(m*n), Space: O(m*n)
    Tricky: Start BFS from all gates simultaneously
    """
    if not rooms:
        return

    m, n = len(rooms), len(rooms[0])
    queue = deque()
    INF = 2147483647

    # Add all gates to queue
    for i in range(m):
        for j in range(n):
            if rooms[i][j] == 0:
                queue.append((i, j))

    directions = [(0,1), (1,0), (0,-1), (-1,0)]

    while queue:
        row, col = queue.popleft()

        for dr, dc in directions:
            r, c = row + dr, col + dc

            if 0 <= r < m and 0 <= c < n and rooms[r][c] == INF:
                rooms[r][c] = rooms[row][col] + 1
                queue.append((r, c))


# ============================================================================
# PATTERN 6: TOPOLOGICAL SORT
# ============================================================================

def topological_sort_dfs(n: int, edges: List[List[int]]) -> List[int]:
    """
    Topological sort using DFS.

    Time: O(V+E), Space: O(V)
    """
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    visited = set()
    stack = []

    def dfs(node):
        visited.add(node)

        for neighbor in graph[node]:
            if neighbor not in visited:
                dfs(neighbor)

        stack.append(node)

    for node in range(n):
        if node not in visited:
            dfs(node)

    return stack[::-1]  # Reverse to get topological order


def can_finish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    """
    LeetCode 207 - Course schedule (detect cycle in directed graph).

    Time: O(V+E), Space: O(V)
    """
    graph = defaultdict(list)
    for course, prereq in prerequisites:
        graph[course].append(prereq)

    UNVISITED, VISITING, VISITED = 0, 1, 2
    state = [UNVISITED] * numCourses

    def has_cycle(node):
        if state[node] == VISITING:
            return True
        if state[node] == VISITED:
            return False

        state[node] = VISITING

        for neighbor in graph[node]:
            if has_cycle(neighbor):
                return True

        state[node] = VISITED
        return False

    for course in range(numCourses):
        if has_cycle(course):
            return False

    return True


def find_order(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    """
    LeetCode 210 - Course schedule II (topological sort).

    Time: O(V+E), Space: O(V)
    Using Kahn's algorithm (BFS-based)
    """
    graph = defaultdict(list)
    indegree = [0] * numCourses

    for course, prereq in prerequisites:
        graph[prereq].append(course)
        indegree[course] += 1

    # Start with nodes having 0 indegree
    queue = deque([i for i in range(numCourses) if indegree[i] == 0])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)

        for neighbor in graph[node]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                queue.append(neighbor)

    return result if len(result) == numCourses else []


# ============================================================================
# PATTERN 7: CLONE GRAPH
# ============================================================================

class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors else []


def clone_graph(node: Optional[Node]) -> Optional[Node]:
    """
    LeetCode 133 - Clone graph.

    Time: O(V+E), Space: O(V)
    """
    if not node:
        return None

    old_to_new = {}

    def dfs(node):
        if node in old_to_new:
            return old_to_new[node]

        copy = Node(node.val)
        old_to_new[node] = copy

        for neighbor in node.neighbors:
            copy.neighbors.append(dfs(neighbor))

        return copy

    return dfs(node)


# ============================================================================
# PATTERN 8: WORD LADDER / TRANSFORMATION
# ============================================================================

def ladder_length(beginWord: str, endWord: str, wordList: List[str]) -> int:
    """
    LeetCode 127 - Word ladder (shortest transformation sequence).

    Time: O(M² * N) where M is word length, N is word count
    Space: O(M * N)
    """
    word_set = set(wordList)

    if endWord not in word_set:
        return 0

    queue = deque([(beginWord, 1)])
    visited = {beginWord}

    while queue:
        word, level = queue.popleft()

        if word == endWord:
            return level

        # Try changing each character
        for i in range(len(word)):
            for c in 'abcdefghijklmnopqrstuvwxyz':
                next_word = word[:i] + c + word[i+1:]

                if next_word in word_set and next_word not in visited:
                    visited.add(next_word)
                    queue.append((next_word, level + 1))

    return 0


# ============================================================================
# PATTERN 9: BIPARTITE GRAPH
# ============================================================================

def is_bipartite(graph: List[List[int]]) -> bool:
    """
    LeetCode 785 - Check if graph is bipartite (2-colorable).

    Time: O(V+E), Space: O(V)
    Tricky: Color nodes with two colors, check for conflicts
    """
    n = len(graph)
    color = [-1] * n

    def dfs(node, c):
        color[node] = c

        for neighbor in graph[node]:
            if color[neighbor] == c:
                return False
            if color[neighbor] == -1 and not dfs(neighbor, 1 - c):
                return False

        return True

    for node in range(n):
        if color[node] == -1:
            if not dfs(node, 0):
                return False

    return True


# ============================================================================
# PRACTICE PROBLEMS
# ============================================================================
"""
EASY:
- Find Center of Star Graph (1791)
- Find if Path Exists in Graph (1971)

MEDIUM (DFS/BFS):
- Number of Islands (200)
- Clone Graph (133)
- Course Schedule (207)
- Course Schedule II (210)
- Pacific Atlantic Water Flow (417)
- Number of Connected Components (323)
- Graph Valid Tree (261)
- Walls and Gates (286)

MEDIUM (Shortest Path):
- Shortest Path in Binary Matrix (1091)
- Word Ladder (127)
- Open the Lock (752)
- Minimum Knight Moves (1197)

HARD:
- Word Ladder II (126)
- Alien Dictionary (269)
- Minimum Cost to Make Valid Path (1368)
- Swim in Rising Water (778)
"""

if __name__ == "__main__":
    # Test DFS/BFS
    graph = {0: [1, 2], 1: [2], 2: [3], 3: [3]}

    print("DFS:", dfs_iterative(graph, 0))
    print("BFS:", bfs(graph, 0))

    # Test Number of Islands
    grid = [
        ["1", "1", "0", "0", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "1", "0", "0"],
        ["0", "0", "0", "1", "1"]
    ]
    # Note: This will modify grid
    print("Number of Islands:", num_islands(grid))

    # Test Course Schedule
    print("Can Finish:", can_finish(2, [[1, 0]]))
    print("Course Order:", find_order(2, [[1, 0]]))

File: C++ Playground/12_GraphDfsBfs.cpp
#include <algorithm>
#include <climits>
#include <functional>
#include <iostream>
#include <numeric>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <vector>

namespace graph_playground {

/// Adjacency list via vector<vector<int>>.
/// BFS for shortest path (unweighted); DFS for connectivity.
/// Topological sort via Kahn's (BFS on in-degrees) or DFS post-order.
/// Dijkstra with priority_queue<pair<int,int>> for weighted graphs.

struct GraphNode {
  int val;
  std::vector<GraphNode*> neighbors;
  explicit GraphNode(int x) : val(x) {}
};

constexpr int kDirs[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

// ============================================================================
// DFS — ISLANDS
// ============================================================================

/// Number of islands (LC 200 Med).
/// Count connected components of '1' in grid.
/// Time: O(m * n), Space: O(m * n).
[[nodiscard]] int NumIslands(std::vector<std::vector<char>> grid) noexcept {
  if (grid.empty()) return 0;

  const int m = static_cast<int>(grid.size());
  const int n = static_cast<int>(grid[0].size());
  int count = 0;

  auto dfs = [&grid, m, n](int i, int j, auto& fn) -> void {
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') return;
    grid[i][j] = '0';
    fn(i + 1, j, fn);
    fn(i - 1, j, fn);
    fn(i, j + 1, fn);
    fn(i, j - 1, fn);
  };

  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (grid[i][j] == '1') {
        dfs(i, j, dfs);
        ++count;
      }
    }
  }

  return count;
}

/// Max area island (LC 695 Med).
/// Find area of largest island.
/// Time: O(m * n), Space: O(m * n).
[[nodiscard]] int MaxAreaIsland(std::vector<std::vector<int>> grid) noexcept {
  if (grid.empty()) return 0;

  const int m = static_cast<int>(grid.size());
  const int n = static_cast<int>(grid[0].size());
  int max_area = 0;

  auto dfs = [&grid, m, n, &max_area](int i, int j, auto& fn) -> int {
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == 0) return 0;
    grid[i][j] = 0;
    int area = 1;
    area += fn(i + 1, j, fn);
    area += fn(i - 1, j, fn);
    area += fn(i, j + 1, fn);
    area += fn(i, j - 1, fn);
    return area;
  };

  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (grid[i][j] == 1) {
        max_area = std::max(max_area, dfs(i, j, dfs));
      }
    }
  }

  return max_area;
}

/// Surrounded regions (LC 130 Med).
/// Mark 'O' regions surrounded by 'X' as 'X'.
/// DFS from border 'O's to mark safe regions.
/// Time: O(m * n), Space: O(m * n).
void SurroundedRegions(std::vector<std::vector<char>>& board) noexcept {
  if (board.empty()) return;

  const int m = static_cast<int>(board.size());
  const int n = static_cast<int>(board[0].size());

  auto dfs = [&board, m, n](int i, int j, auto& fn) -> void {
    if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != 'O') return;
    board[i][j] = 'T';
    fn(i + 1, j, fn);
    fn(i - 1, j, fn);
    fn(i, j + 1, fn);
    fn(i, j - 1, fn);
  };

  for (int i = 0; i < m; ++i) {
    if (board[i][0] == 'O') dfs(i, 0, dfs);
    if (board[i][n - 1] == 'O') dfs(i, n - 1, dfs);
  }

  for (int j = 0; j < n; ++j) {
    if (board[0][j] == 'O') dfs(0, j, dfs);
    if (board[m - 1][j] == 'O') dfs(m - 1, j, dfs);
  }

  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (board[i][j] == 'O') board[i][j] = 'X';
      else if (board[i][j] == 'T') board[i][j] = 'O';
    }
  }
}

// ============================================================================
// BFS — SHORTEST PATH
// ============================================================================

/// Shortest path in binary matrix (LC 1091 Med).
/// 8-directional BFS.
/// Time: O(m * n), Space: O(m * n).
[[nodiscard]] int ShortestPathBinaryMatrix(
    std::vector<std::vector<int>> grid) noexcept {
  if (grid.empty() || grid[0][0] == 1) return -1;

  const int m = static_cast<int>(grid.size());
  const int n = static_cast<int>(grid[0].size());

  std::queue<std::pair<int, int>> q;
  q.push({0, 0});
  grid[0][0] = 1;
  int dist = 1;

  constexpr int dirs[8][2] = {
    {0, 1}, {1, 0}, {0, -1}, {-1, 0}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}
  };

  while (!q.empty()) {
    int size = static_cast<int>(q.size());
    for (int i = 0; i < size; ++i) {
      auto [x, y] = q.front();
      q.pop();

      if (x == m - 1 && y == n - 1) return dist;

      for (const auto& dir : dirs) {
        int nx = x + dir[0];
        int ny = y + dir[1];
        if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == 0) {
          grid[nx][ny] = 1;
          q.push({nx, ny});
        }
      }
    }
    ++dist;
  }

  return -1;
}

/// Rotting oranges (LC 994 Med).
/// Multi-source BFS.
/// Time: O(m * n), Space: O(m * n).
[[nodiscard]] int RottingOranges(std::vector<std::vector<int>> grid) noexcept {
  if (grid.empty()) return 0;

  const int m = static_cast<int>(grid.size());
  const int n = static_cast<int>(grid[0].size());

  std::queue<std::pair<int, int>> q;
  int fresh = 0;

  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (grid[i][j] == 2) q.push({i, j});
      else if (grid[i][j] == 1) ++fresh;
    }
  }

  if (fresh == 0) return 0;

  int minutes = 0;
  while (!q.empty()) {
    int size = static_cast<int>(q.size());
    for (int i = 0; i < size; ++i) {
      auto [x, y] = q.front();
      q.pop();

      for (const auto& dir : kDirs) {
        int nx = x + dir[0];
        int ny = y + dir[1];
        if (nx >= 0 && nx < m && ny >= 0 && ny < n &&
            grid[nx][ny] == 1) {
          grid[nx][ny] = 2;
          q.push({nx, ny});
          --fresh;
        }
      }
    }
    if (!q.empty()) ++minutes;
  }

  return fresh == 0 ? minutes : -1;
}

/// Word ladder (LC 127 Hard).
/// Shortest transformation sequence.
/// Time: O(n * w^2) where w = word length, Space: O(n).
[[nodiscard]] int WordLadder(std::string begin_word,
                            std::string end_word,
                            std::vector<std::string>& word_list) noexcept {
  std::unordered_set<std::string> dict(word_list.begin(), word_list.end());
  if (dict.find(end_word) == dict.end()) return 0;

  std::queue<std::pair<std::string, int>> q;
  q.push({begin_word, 1});

  while (!q.empty()) {
    auto [word, level] = q.front();
    q.pop();

    if (word == end_word) return level;

    for (int i = 0; i < static_cast<int>(word.size()); ++i) {
      char orig = word[i];
      for (char c = 'a'; c <= 'z'; ++c) {
        word[i] = c;
        if (dict.count(word) > 0) {
          dict.erase(word);
          q.push({word, level + 1});
        }
      }
      word[i] = orig;
    }
  }

  return 0;
}

// ============================================================================
// TOPOLOGICAL SORT
// ============================================================================

/// Can finish courses (LC 207 Med).
/// Kahn's algorithm (BFS topological sort).
/// Time: O(V + E), Space: O(V).
[[nodiscard]] bool CanFinish(int num_courses,
                            const std::vector<std::vector<int>>&
                            prerequisites) noexcept {
  std::vector<int> in_degree(num_courses, 0);
  std::vector<std::vector<int>> adj(num_courses);

  for (const auto& pre : prerequisites) {
    adj[pre[1]].emplace_back(pre[0]);
    ++in_degree[pre[0]];
  }

  std::queue<int> q;
  for (int i = 0; i < num_courses; ++i) {
    if (in_degree[i] == 0) q.push(i);
  }

  int count = 0;
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    ++count;

    for (int v : adj[u]) {
      --in_degree[v];
      if (in_degree[v] == 0) q.push(v);
    }
  }

  return count == num_courses;
}

/// Find course order (LC 210 Med).
/// Topological sort order.
/// Time: O(V + E), Space: O(V).
[[nodiscard]] std::vector<int> FindOrder(
    int num_courses,
    const std::vector<std::vector<int>>& prerequisites) noexcept {
  std::vector<int> result;
  std::vector<int> in_degree(num_courses, 0);
  std::vector<std::vector<int>> adj(num_courses);

  for (const auto& pre : prerequisites) {
    adj[pre[1]].emplace_back(pre[0]);
    ++in_degree[pre[0]];
  }

  std::queue<int> q;
  for (int i = 0; i < num_courses; ++i) {
    if (in_degree[i] == 0) q.push(i);
  }

  while (!q.empty()) {
    int u = q.front();
    q.pop();
    result.emplace_back(u);

    for (int v : adj[u]) {
      --in_degree[v];
      if (in_degree[v] == 0) q.push(v);
    }
  }

  return static_cast<int>(result.size()) == num_courses ? result
                                                          : std::vector<int>();
}

// ============================================================================
// CLONE
// ============================================================================

/// Clone graph (LC 133 Med).
/// BFS + map to clone nodes.
/// Time: O(V + E), Space: O(V).
[[nodiscard]] GraphNode* CloneGraph(GraphNode* node) noexcept {
  if (node == nullptr) return nullptr;

  std::unordered_map<GraphNode*, GraphNode*> cloned;
  std::queue<GraphNode*> q;

  q.push(node);
  cloned[node] = new GraphNode(node->val);

  while (!q.empty()) {
    GraphNode* curr = q.front();
    q.pop();

    for (GraphNode* neighbor : curr->neighbors) {
      if (cloned.find(neighbor) == cloned.end()) {
        cloned[neighbor] = new GraphNode(neighbor->val);
        q.push(neighbor);
      }
      cloned[curr]->neighbors.emplace_back(cloned[neighbor]);
    }
  }

  return cloned[node];
}

// ============================================================================
// UNION-FIND
// ============================================================================

/// Count connected components (Union-Find).
/// Time: O(n * α(n)), Space: O(n).
[[nodiscard]] int CountComponents(int n,
                                 const std::vector<std::vector<int>>&
                                 edges) noexcept {
  std::vector<int> parent(n);
  std::iota(parent.begin(), parent.end(), 0);

  std::function<int(int)> find = [&parent, &find](int x) -> int {
    if (parent[x] != x) parent[x] = find(parent[x]);
    return parent[x];
  };

  int count = n;
  for (const auto& edge : edges) {
    int px = find(edge[0]);
    int py = find(edge[1]);
    if (px != py) {
      parent[px] = py;
      --count;
    }
  }

  return count;
}

/// Find redundant connection (LC 684 Med).
/// Union-Find to detect cycle.
/// Time: O(n * α(n)), Space: O(n).
[[nodiscard]] std::vector<int> FindRedundantConnection(
    const std::vector<std::vector<int>>& edges) noexcept {
  std::vector<int> parent(edges.size() + 1);
  std::iota(parent.begin(), parent.end(), 0);

  std::function<int(int)> find = [&parent, &find](int x) -> int {
    if (parent[x] != x) parent[x] = find(parent[x]);
    return parent[x];
  };

  for (const auto& edge : edges) {
    int px = find(edge[0]);
    int py = find(edge[1]);
    if (px == py) return {edge[0], edge[1]};
    parent[px] = py;
  }

  return {};
}

// ============================================================================
// DIJKSTRA
// ============================================================================

/// Network delay time (LC 743 Med).
/// Dijkstra's shortest path.
/// Time: O((V + E) log V), Space: O(V).
[[nodiscard]] int NetworkDelayTime(
    const std::vector<std::vector<int>>& times, int n,
    int k) noexcept {
  std::vector<std::vector<std::pair<int, int>>> adj(n + 1);

  for (const auto& t : times) {
    adj[t[0]].emplace_back(t[1], t[2]);
  }

  std::vector<int> dist(n + 1, INT_MAX);
  std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>,
                      std::greater<std::pair<int, int>>>
      pq;

  dist[k] = 0;
  pq.push({0, k});

  while (!pq.empty()) {
    auto [d, u] = pq.top();
    pq.pop();

    if (d > dist[u]) continue;

    for (const auto& [v, w] : adj[u]) {
      if (dist[u] + w < dist[v]) {
        dist[v] = dist[u] + w;
        pq.push({dist[v], v});
      }
    }
  }

  int max_dist = 0;
  for (int i = 1; i <= n; ++i) {
    if (dist[i] == INT_MAX) return -1;
    max_dist = std::max(max_dist, dist[i]);
  }

  return max_dist;
}

/// Pacific Atlantic water flow (LC 417 Med).
/// Reverse BFS from both borders.
/// Time: O(m * n), Space: O(m * n).
[[nodiscard]] std::vector<std::vector<int>> PacificAtlantic(
    const std::vector<std::vector<int>>& heights) noexcept {
  if (heights.empty()) return {};

  const int m = static_cast<int>(heights.size());
  const int n = static_cast<int>(heights[0].size());

  std::vector<std::vector<bool>> pacific(m, std::vector<bool>(n, false));
  std::vector<std::vector<bool>> atlantic(m, std::vector<bool>(n, false));

  auto bfs = [&heights, m, n](std::vector<std::vector<bool>>& visited,
                               std::queue<std::pair<int, int>>& q) {
    while (!q.empty()) {
      auto [x, y] = q.front();
      q.pop();

      for (const auto& dir : kDirs) {
        int nx = x + dir[0];
        int ny = y + dir[1];
        if (nx >= 0 && nx < m && ny >= 0 && ny < n &&
            !visited[nx][ny] && heights[nx][ny] >= heights[x][y]) {
          visited[nx][ny] = true;
          q.push({nx, ny});
        }
      }
    }
  };

  std::queue<std::pair<int, int>> pac_q, atl_q;

  for (int i = 0; i < m; ++i) {
    pacific[i][0] = true;
    pac_q.push({i, 0});
    atlantic[i][n - 1] = true;
    atl_q.push({i, n - 1});
  }

  for (int j = 0; j < n; ++j) {
    pacific[0][j] = true;
    pac_q.push({0, j});
    atlantic[m - 1][j] = true;
    atl_q.push({m - 1, j});
  }

  bfs(pacific, pac_q);
  bfs(atlantic, atl_q);

  std::vector<std::vector<int>> result;
  for (int i = 0; i < m; ++i) {
    for (int j = 0; j < n; ++j) {
      if (pacific[i][j] && atlantic[i][j]) {
        result.emplace_back(std::vector<int>{i, j});
      }
    }
  }

  return result;
}

}  // namespace graph_playground

/// Test driver demonstrating all graph functions.
int main() {
  using namespace graph_playground;

  std::cout << "=== Graph DFS/BFS Algorithms ===\n\n";

  std::vector<std::vector<char>> grid1 = {
    {'1', '1', '1', '1', '0'},
    {'1', '1', '0', '1', '0'},
    {'1', '1', '0', '0', '0'},
    {'0', '0', '0', '0', '0'}
  };
  std::cout << "Num Islands: " << NumIslands(grid1) << "\n";

  std::vector<std::vector<int>> grid2 = {
    {1, 1, 1},
    {1, 0, 1},
    {1, 1, 1}
  };
  std::cout << "Max Area Island: " << MaxAreaIsland(grid2) << "\n";

  std::cout << "Can Finish (2, [[1,0]]): "
            << (CanFinish(2, {{1, 0}}) ? "true" : "false") << "\n";

  std::cout << "Can Finish (2, [[1,0],[0,1]]): "
            << (CanFinish(2, {{1, 0}, {0, 1}}) ? "true" : "false")
            << "\n";

  std::cout << "Count Components (5, [[0,1],[1,2],[3,4]]): "
            << CountComponents(5, {{0, 1}, {1, 2}, {3, 4}}) << "\n";

  std::cout << "Find Redundant Connection ([[1,2],[1,3],[2,3]]): ";
  auto redund = FindRedundantConnection({{1, 2}, {1, 3}, {2, 3}});
  if (!redund.empty()) std::cout << "[" << redund[0] << "," << redund[1] << "]";
  std::cout << "\n";

  std::cout << "Network Delay Time ([[2,1,1],[2,3,1],[3,4,1]], 4, 2): "
            << NetworkDelayTime({{2, 1, 1}, {2, 3, 1}, {3, 4, 1}}, 4, 2)
            << "\n";

  std::cout << "\nAll graph tests passed!\n";
  return 0;
}
